README
DjangoPyPI
==========

DjangoPyPI is a Django application that provides a re-implementation of the 
`Python Package Index <http://pypi.python.org>`_.  

Installation
------------

Path
____

The first step is to get ``djangopypi`` into your Python path.

Buildout
++++++++

Simply add ``djangopypi`` to your list of ``eggs`` and run buildout again it 
should downloaded and installed properly.

EasyInstall/Setuptools
++++++++++++++++++++++

If you have setuptools installed, you can use ``easy_install djangopypi``

Manual
++++++

Download and unpack the source then run::

    $ python setup.py install

Django Settings
_______________

Add ``djangopypi`` to your ``INSTALLED_APPS`` setting and run ``syncdb`` again 
to get the database tables [#]_.

Then add an include in your url config for ``djangopypi.urls``::

    urlpatterns = patterns("",
        ...
        url(r'', include("djangopypi.urls"))
    )

This will make the repository interface be accessible at ``/pypi/``.



Uploading to your PyPI
----------------------

Assuming you are running your Django site locally for now, add the following to 
your ``~/.pypirc`` file::

    [distutils]
    index-servers =
        pypi
        local

    [pypi]
    username:user
    password:secret

    [local]
    username:user
    password:secret
    repository:http://localhost:8000/pypi

Uploading a package
___________________

Before a user can upload a package they must be granted permission to do
so. In the admin interface assign the user the djangopypi.add_package
permission.

Failure to do so will give a 403 when uploading

Uploading a package: Python >=2.6
_________________________________

To push the package to the local pypi::

    $ python setup.py register -r local sdist upload -r local


Uploading a package: Python <2.6
________________________________

If you don't have Python 2.6 please run the command below to install the 
backport of the extension for multiple repositories::

     $ easy_install -U collective.dist

Instead of using register and dist command, you can use ``mregister`` and 
``mupload`` which are a backport of python 2.6 register and upload commands 
that supports multiple servers.

To push the package to the local pypi::

    $ python setup.py mregister -r local sdist mupload -r local

.. [#] ``djangopypi`` is South enabled, if you are using South then you will need
   to run the South ``migrate`` command to get the tables.


Changelog
History
=======

0.4.4-isotoma21-ampledata1 (2012-08-17)
------------------
* Re-enabled RSS URL. @ampledata

0.4.3 (2011-02-22)
------------------

* Moved xmlrpc views into views folder
* Moved xmlrpc command settings to the settings file
* Cleaned up xmlrpc views to remove django.contrib.sites dependency

0.4.2 (2011-02-21)
------------------

* Added CSRF support for Django>=1.2
* Added conditional support to proxy packages not indexed

0.4.1 (2010-06-17)
------------------

* Added conditional support for django-haystack searching

0.4 (2010-06-14)
----------------

* 'list_classifiers' action handler
* Issue #3: decorators imports incompatible with Django 1.0, 1.1
* RSS support for release index, packages
* Distribution uploads (files for releases)

0.3.1 (2010-06-09)
------------------

* Installation bugfix

0.3 (2010-06-09)
----------------

* Added DOAP views of packages and releases
* Splitting djangopypi off of chishop
* Switched most views to using django generic views

Backwards incompatible changes
______________________________

* Refactored package/project model to support multiple owners/maintainers
* Refactored release to match the metadata only that exists on pypi.python.org
* Created a Distribution model for distribution files on a release

0.2.0 (2009-03-22)
------------------

* Registering projects and uploading releases now requires authentication.
* Every project now has an owner, so only the user registering the project can 
  add releases.
* md5sum is now properly listed in the release link.
* Project names can now have dots ('.') in them.
* Fixed a bug where filenames was mangled if the distribution file already existed.
* Releases now list both project name and version, instead of just version in the admin interface.
* Added a sample buildout.cfg. Thanks to Rune Halvorsen (runeh@opera.com).

Backwards incompatible changes
______________________________

* Projects now has an associated owner, so old projects must be exported and 
  imported to a new database.

0.1.0 (2009-03-22)
------------------

* Initial release


djangopypi/settings.py
from django.conf import settings
import os
import chishop

# This is disabled on pypi.python.org, can be useful if you make mistakes
if not hasattr(settings,'DJANGOPYPI_ALLOW_VERSION_OVERWRITE'):
    settings.DJANGOPYPI_ALLOW_VERSION_OVERWRITE = False

""" The upload_to argument for the file field in releases. This can either be 
a string for a path relative to your media folder or a callable. For more 
information, see http://docs.djangoproject.com/ """
if not hasattr(settings,'DJANGOPYPI_RELEASE_UPLOAD_TO'):
    settings.DJANGOPYPI_RELEASE_UPLOAD_TO = 'dists'

if not hasattr(settings, 'DJANGOPYPI_RELEASE_URL'):
    settings.DJANGOPYPI_RELEASE_URL = '/packages/'

if not hasattr(settings,'DJANGOPYPI_OS_NAMES'):
    settings.DJANGOPYPI_OS_NAMES = (
        ("aix", "AIX"),
        ("beos", "BeOS"),
        ("debian", "Debian Linux"),
        ("dos", "DOS"),
        ("freebsd", "FreeBSD"),
        ("hpux", "HP/UX"),
        ("mac", "Mac System x."),
        ("macos", "MacOS X"),
        ("mandrake", "Mandrake Linux"),
        ("netbsd", "NetBSD"),
        ("openbsd", "OpenBSD"),
        ("qnx", "QNX"),
        ("redhat", "RedHat Linux"),
        ("solaris", "SUN Solaris"),
        ("suse", "SuSE Linux"),
        ("yellowdog", "Yellow Dog Linux"),
    )

if not hasattr(settings,'DJANGOPYPI_ARCHITECTURES'):
    settings.DJANGOPYPI_ARCHITECTURES = (
        ("alpha", "Alpha"),
        ("hppa", "HPPA"),
        ("ix86", "Intel"),
        ("powerpc", "PowerPC"),
        ("sparc", "Sparc"),
        ("ultrasparc", "UltraSparc"),
    )

if not hasattr(settings,'DJANGOPYPI_DIST_FILE_TYPES'):
    settings.DJANGOPYPI_DIST_FILE_TYPES = (
        ('sdist','Source'),
        ('bdist_dumb','"dumb" binary'),
        ('bdist_rpm','RPM'),
        ('bdist_wininst','MS Windows installer'),
        ('bdist_egg','Python Egg'),
        ('bdist_dmg','OS X Disk Image'),
    )

if not hasattr(settings,'DJANGOPYPI_PYTHON_VERSIONS'):
    settings.DJANGOPYPI_PYTHON_VERSIONS = (
        ('any','Any i.e. pure python'),
        ('2.1','2.1'),
        ('2.2','2.2'),
        ('2.3','2.3'),
        ('2.4','2.4'),
        ('2.5','2.5'),
        ('2.6','2.6'),
        ('2.7','2.7'),
        ('3.0','3.0'),
        ('3.1','3.1'),
        ('3.2','3.2'),
    )

if not hasattr(settings, 'DJANGOPYPI_METADATA_FIELDS'):
    settings.DJANGOPYPI_METADATA_FIELDS = {
        '1.0': ('platform','summary','description','keywords','home_page',
                'author','author_email', 'license', ),
        '1.1': ('platform','supported_platform','summary','description',
                'keywords','home_page','download_url','author','author_email',
                'license','classifier','requires','provides','obsoletes',),
        '1.2': ('platform','supported_platform','summary','description',
                'keywords','home_page','download_url','author','author_email',
                'maintainer','maintainer_email','license','classifier',
                'requires_dist','provides_dist','obsoletes_dist',
                'requires_python','requires_external','project_url')}

if not hasattr(settings, 'DJANGOPYPI_METADATA_FORMS'):
    from djangopypi.forms import Metadata10Form, Metadata11Form, Metadata12Form
    settings.DJANGOPYPI_METADATA_FORMS = {
        '1.0': Metadata10Form,
        '1.1': Metadata11Form,
        '1.2': Metadata12Form}

if not hasattr(settings, 'DJANGOPYPI_FALLBACK_VIEW'):
    from djangopypi.views import releases 
    settings.DJANGOPYPI_FALLBACK_VIEW = releases.index

if not hasattr(settings,'DJANGOPYPI_ACTION_VIEWS'):
    from djangopypi.views import distutils
    
    settings.DJANGOPYPI_ACTION_VIEWS = {
        "file_upload": distutils.register_or_upload, #``sdist`` command
        "submit": distutils.register_or_upload, #``register`` command
        "list_classifiers": distutils.list_classifiers, #``list_classifiers`` command
    }

if not hasattr(settings,'DJANGOPYPI_XMLRPC_COMMANDS'):
    from djangopypi.views import xmlrpc
    
    settings.DJANGOPYPI_XMLRPC_COMMANDS = {
        'list_packages': xmlrpc.list_packages,
        'package_releases': xmlrpc.package_releases,
        'release_urls': xmlrpc.release_urls,
        'release_data': xmlrpc.release_data,
        #'search': xmlrpc.search, Not done yet
        #'changelog': xmlrpc.changelog, Not done yet
        #'ratings': xmlrpc.ratings, Not done yet
    }

""" These settings enable proxying of packages that are not in the local index 
to another index, http://pypi.python.org/ by default. This feature is disabled 
by default and can be enabled by setting DJANGOPYPI_PROXY_MISSING to True in 
your settings file. """
if not hasattr(settings, 'DJANGOPYPI_PROXY_BASE_URL'):
    settings.DJANGOPYPI_PROXY_BASE_URL = 'http://pypi.python.org/simple'

if not hasattr(settings, 'DJANGOPYPI_PROXY_MISSING'):
    settings.DJANGOPYPI_PROXY_MISSING = False


""" If deployment doesnt opt to use a nice sensible sendfile implementation,
at least use the simple one """
if not hasattr(settings, 'SENDFILE_BACKEND'):
    settings.SENDFILE_BACKEND = 'sendfile.backends.simple'

